Models/GitHistoryModel.cpp

```diff
@@ -0,0 +1,83 @@
+#include "GitHistoryModel.h"
+#include "Components/GitTreeItem.h"
+
+#include <QDebug>
+
+GitHistoryModel::GitHistoryModel(QObject* parent) : QAbstractTableModel(parent) {}
+
+GitHistoryModel::~GitHistoryModel() {
+  for (git_commit* c : _commits)
+    git_commit_free(c);
+}
+
+void GitHistoryModel::LoadRepo(git_repository *repo) {
+  beginResetModel();
+
+  for (git_commit* c : _commits)
+    git_commit_free(c);
+
+  _commits.clear();
+
+  git_revwalk *walker;
+
+  int error = git_revwalk_new(&walker, repo);
+  if (error != 0) emit GitError();
+
+  error = git_revwalk_push_head(walker);
+  if (error != 0) emit GitError();
+
+  error = git_revwalk_push_ref(walker, "HEAD");
+  if (error != 0) emit GitError();
+
+  git_revwalk_sorting(walker, GIT_SORT_NONE);
+
+  git_oid oid;
+  while (!git_revwalk_next(&oid, walker)) {
+    git_commit* commit = nullptr;
+    git_commit_lookup(&commit, repo, &oid);
+    _commits.append(commit);
+  }
+
+  git_revwalk_free(walker);
+
+  endResetModel();
+}
+
+int GitHistoryModel::rowCount(const QModelIndex &parent) const {
+  if (parent.isValid()) return 0;
+  return _commits.count();
+}
+
+int GitHistoryModel::columnCount(const QModelIndex &parent) const {
+  if (parent.isValid()) return 0;
+  return 5;
+}
+
+QVariant GitHistoryModel::headerData(int section, Qt::Orientation orientation, int role) const {
+  if (role != Qt::DisplayRole || orientation != Qt::Orientation::Horizontal) return QVariant();
+
+  switch(section) {
+    case 0: return tr("Tree");
```

Comment:
I don't know a better way of doing this, but I feel like there probably is one... I mean, naively, you can just put those labels in an array. But I'm not sure what happens if the user tries to drag columns around or something; I feel like Qt probably has better support for this somewhere.